node/test/Infrastructure/Generator.hs

```diff
@@ -31,22 +67,213 @@ genChainUsingModel GeneratorModel{..} =
+    -- agent.
+    --
+    -- TODO: we need to assert that the '0 < totalStake'.
+    un :: Int
+    (un, _) = seeds s0 totalStake !! n
+
+    -- | Note that we should be able to find an index @i@ such that @un <
+    -- cStake !! i@, since @un < cStake !! (length cStake - 1) == totalStake@.
+    --
+    -- TODO: consider using liquid Haskell to actually verify this!
+    err =  error $  "The impossible has happened: "
+                 <> "could not find an index that accumulates enough stake."
+
+-- | Sequence of random number and seeds.
+seeds :: Seed -> Int -> [(Int, Seed)]
+seeds _ 0  = []
+seeds s0 n = (n, s0):[ numGen n sn | (_, sn) <- seeds s0 n ]
+
+-- | Implementation of the number generator function as described in Section
+-- 7.2 (called @f@).
+--
+-- Here the 'Int' values represent stake values.
+--
+-- Ideally `numGen` should be surjective so that each stakeholder could be
+-- selected by some random seed.
+--
+-- Note that this function is not defined when @i == 0@
+numGen :: Int  -> Seed -> (Int, Seed)
+-- Some dummy implementation for now...
+numGen i (Seed s) = (s `mod` i, Seed (s + 1))
+
+-- | For now we return the same seed. Ideally we'd need to perform an SCC
+-- computation, which uses the stake distribution and the VSS payloads (which
+-- should be tracked in the state).
+--
+mCurrentSeed :: TransState GivenHash -> Seed
+mCurrentSeed = const (Seed 15)
+
+-- | Compute the number of blocks as the height of the chain (See Section 7.1).
+-- The number of blocks can be obtained from the list of checkpoints in the
+-- translation state, which contains a checkpoint for each block that we
+-- translate.
+mHeight :: TransState GivenHash -> Int
+mHeight st = st ^. tsCheckpoints . to length
+
+-- | Quality of a chain. It maps a number of blocks into a number of blocks.
+mQuality :: Int -> Int
+mQuality = (*2)
+
+-- | Is the slot in the commitment phase?
+mInCommitmentPhase :: SlotId -> Bool
+mInCommitmentPhase (SlotId k) = 0 <= k && k < mQuality k
+
+-- | Is the slot in the opening phase.
+mInOpenPhase :: SlotId -> Bool
+mInOpenPhase (SlotId k) = 2 * mQuality k <= k && k < 3 * mQuality k
+
+-- | Is the slot in the recovery phase.
+mInRecoveryPhase :: SlotId -> Bool
+mInRecoveryPhase (SlotId k) = 4 * mQuality k <= k && k < 5 * mQuality k
+
+-- | Initial stake distribution that assigns 1 to each address passed as
+-- parameter.
+--
+-- TODO: shouldn't this be determined based on other parameters?
+mInitialStakeDistribution :: NonEmpty Addr -> StakeDistribution Addr
+mInitialStakeDistribution addrs = StakeDistribution
+                                $ Map.fromList
+                                $ zip (toList addrs) (repeat 1)
+
+-- | Extract the set of addresses in a chain.
+chainAddresses :: DSL.Chain h a -> [a]
+chainAddresses = map DSL.outAddr
+               . concatMap DSL.trOuts
+               . concatMap toList
+               . toList
+
+-- | Translation of DSL transactions onto Abstract transactions
```

Comment:
Functions for translating DSL types into Abstract types already exist in the `FromUTxO` module. They should be reused, or if their signatures are not suitable, then they should be changed so they fit both purposes.

Reply:
There is, as part of the `translate` function and written as a where clause function. So it should be made a top-level function.